recipes/full_finetune_single_device.py

```diff
@@ -433,73 +438,44 @@ def _setup_optimizer(
-            if opt_state_dict is not None:
-                try:
-                    self._optim_ckpt_wrapper.load_state_dict(opt_state_dict)
-                except BaseException as e:
-                    raise RuntimeError(
-                        "Failed loading in-backward optimizer checkpoints."
-                        "Please make sure run being restored from was using in-backward optimizer."
-                    ) from e
-            self._logger.info("In-backward optimizers are set up.")
-            return None
         else:
             optimizer = config.instantiate(cfg_optimizer, self._model.parameters())
 
-            if opt_state_dict:
-                optimizer.load_state_dict(opt_state_dict)
-            self._logger.info("Optimizer is initialized.")
-            return optimizer
+        if opt_state_dict:
+            optimizer.load_state_dict(opt_state_dict)
+
+        self._logger.info("Optimizer is initialized.")
+        return optimizer
 
     def _setup_lr_scheduler(
         self,
-        cfg_lr_scheduler: Optional[DictConfig],
+        cfg_lr_scheduler: DictConfig,
         num_training_steps: int,
         last_epoch: int,
-    ) -> Optional[Optimizer]:
+    ) -> LambdaLR:
         """
         Set up the learning rate scheduler based on the provided configuration.
         It handles both standard optimization and optimizer-in-backward cases, and supports
         schedulers from both torchtune.modules and torch.optim.
-
-        Args:
-            cfg_lr_scheduler (Optional[DictConfig]): The learning rate scheduler configuration.
-            num_training_steps (int): The total number of training steps.
-            last_epoch (int): The index of the last epoch.
-
-        Returns:
-            lr_scheduler (Optional[Optimizer]): The learning rate scheduler.
         """
-        if cfg_lr_scheduler is None:
-            self._logger.info(
-                "No learning rate scheduler configured. Using constant learning rate."
-            )
-            return None
-
-        if self._optimizer_in_bwd:
+        if isinstance(self._optimizer, OptimizerInBackwardWrapper):
             # Use the first optimizer from the wrapper to represent the learning rate
-            optimizer = next(iter(self._optim_ckpt_wrapper.optim_map.values()))
+            optimizer = next(iter(self._optimizer.optim_map.values()))
         else:
-            # Standard case: use the single optimizer
             optimizer = self._optimizer
 
-        # Instantiate the learning rate scheduler
         lr_scheduler = config.instantiate(
             cfg_lr_scheduler,
             optimizer,
             num_training_steps=num_training_steps,
             last_epoch=last_epoch,
         )
 
-        if self._optimizer_in_bwd:
-            # Modify the scheduler for optimizer_in_bwd case
-            self._optim_ckpt_wrapper.set_lr_scheduler(lr_scheduler)
+        if isinstance(self._optimizer, OptimizerInBackwardWrapper):
```

Comment:
Personally I like this like 5% more but I know @pbontrager hates `isinstance` checks

Reply:
@pbontrager just told me it's fine and that I'm the best torchtune engineer.